README.md
# demo_animation

Flutter动画插件，

## Getting Started

使用方式比较简单，只需设定起始坐标，设置需要改变的空间即可.
```Dart
AudioAllMicGift(
    endOffset:RandomScreenOffset.generateRandomScreenCoordinates(context),
    durationTime: 3000,
    offSetY: 100,
    childWidget: const Icon(Icons.accessibility),
)
```


直接上效果

![gif](/images/tutieshi_558x1024_1s.gif)

#### 希望它能够给你带来帮助！

CHANGELOG.md
# change Log
## [1.0.1] 2023-05-18
天女散花动画！AudioAllMic

test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_animation_factory/animation_demo.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


example/lib/main.dart
import 'package:demo_animation/audio_all_mic_gift.dart';
import 'package:demo_animation/tools/random_screen_offset.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Widget> widgetList = [];
  bool isOnClick = false;
  void _incrementCounter() async {
    if (!isOnClick) {
      isOnClick = true;
      for (int i = 0; i < 100000; i++) {
        await Future.delayed(const Duration(milliseconds: 150)); // 模拟异步操作
        setState(() {
          widgetList.add(AudioAllMicGift(
            endOffset:
                RandomScreenOffset.generateRandomScreenCoordinates(context),
            // childWidget: const Image(
            // image: AssetImage("images/xiayu.gif"), width: 100.0),
            childWidget: const Icon(Icons.accessibility),
          ));
        });
      }
    } else {
      return;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Stack(alignment: Alignment.center, children: widgetList),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/audio_all_mic_gift.dart
import 'package:flutter/material.dart';

import 'animation/animation_alpha_move_alpha.dart';
import 'animation/animation_alpha_scale_move_alpha.dart';

///动画状态回调 [status]动画状态每个阶段状态
typedef AnimationStateCallback = Function(AnimationStatus status);

class AudioAllMicGift extends StatefulWidget {
  //动画状态监听
  final AnimationStateCallback? animationStateCallback;
  //起点坐标 默认屏幕中间点
  final Offset startOffset;
  //终点坐标
  final Offset endOffset;
  //动画控制的空间组件widget
  final Widget? childWidget;
  //y坐标移动的距离 默认50
  final double offSetY;
  //动画持续的时间 默认4500
  final int durationTime;

  ///多轨迹动画
  ///[startOffset] 起始位置，默认参数为屏幕中心点
  ///[childWidget] 动画控制的小组件，widget，
  ///[endOffset]   结束位置，#Offset(dx,dy)
  ///[offSetY]     最后偏移位置的距离 Y轴偏移量
  ///[durationTime]动画持续时间，int 单位毫秒
  ///[animationStateCallback] 动画状态回调，开始，结束，重播回调
  const AudioAllMicGift(
      {super.key,
      this.startOffset = Offset.zero,
      required this.childWidget,
      required this.endOffset,
      this.offSetY = 50,
      this.durationTime = 4500,
      this.animationStateCallback});

  @override
  State<StatefulWidget> createState() {
    return AnimationState();
  }
}

class AnimationState extends State<AudioAllMicGift>
    with TickerProviderStateMixin {
  late AnimationController controller;
  late AnimationController controller2;

  @override
  void initState() {
    super.initState();
    controller = AnimationController(
        vsync: this, duration: Duration(milliseconds: widget.durationTime));
    controller2 = AnimationController(
        vsync: this, duration: Duration(milliseconds: widget.durationTime));

    controller.addStatusListener((status) {
      //动画状态回调
      widget.animationStateCallback?.call(status);
    });

    playAnimation();
  }

  @override
  void dispose() {
    ///释放动画
    controller.reset();
    controller2.reset();
    controller.dispose();
    controller2.dispose();
    super.dispose();
  }

  ///开始播放动画
  playAnimation() {
    //重置动画
    controller.reset();
    controller2.reset();
    //正向播放动画
    controller.forward();
    controller2.forward();
  }

  //创建动画前部分
  createGiftViewAnimation() {
    return AnimationAlphaScaleMoveAlpha(
      controller: controller2,
      startOffset: widget.startOffset,
      endOffset: widget.endOffset,
      offSetY: widget.offSetY,
      childWidget: widget.childWidget,
    );
  }

  //创建动画后部分
  createGiftViewAnimation2() {
    return AnimationAlpahMoveAlpha(
      controller: controller2,
      startOffset: widget.startOffset,
      endOffset: widget.endOffset,
      offSetY: widget.offSetY,
      childWidget: widget.childWidget,
    );
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [createGiftViewAnimation(), createGiftViewAnimation2()],
    );
  }
}


lib/tools/random_screen_offset.dart
import 'dart:math';

import 'package:flutter/material.dart';

class RandomScreenOffset {
  ///获取屏幕随机坐标点，全屏幕
  static Offset generateRandomScreenCoordinates(BuildContext context) {
    Random random = Random();

    // 获取屏幕尺寸
    final screenSize = MediaQuery.of(context).size;
    final screenWidth = screenSize.width;
    final screenHeight = screenSize.height;
    // 计算屏幕中心坐标
    // final centerX = screenWidth / 2;
    // final centerY = screenHeight / 2;
    // // 生成随机 x 坐标（范围：0 到屏幕宽度）
    // double randomX = random.nextDouble() * screenWidth;

    // // 生成随机 y 坐标（范围：0 到屏幕高度）
    // double randomY = random.nextDouble() * screenHeight;

    // 生成随机 x 偏移量（范围：-屏幕宽度/2 到 屏幕宽度/2）
    double randomXOffset =
        (random.nextDouble() * screenWidth) - (screenWidth / 2);

    // 生成随机 y 偏移量（范围：-屏幕高度/2 到 屏幕高度/2）
    double randomYOffset =
        (random.nextDouble() * screenHeight) - (screenHeight / 2);

    // 计算最终随机坐标
    // double randomX = centerX + randomXOffset;
    // double randomY = centerY + randomYOffset;

    return Offset(randomXOffset, randomYOffset);
  }

  static Offset generateRandomScreenCoordinates1_4(BuildContext context) {
    Random random = Random();

    // 获取屏幕尺寸
    final screenSize = MediaQuery.of(context).size;
    final screenWidth = screenSize.width;
    final screenHeight = screenSize.height;
    // 计算屏幕中心坐标
    final centerX = screenWidth / 2;
    final centerY = screenHeight / 2;

    // 生成随机 x 偏移量（范围：-屏幕宽度/2 到 屏幕宽度/2）
    double randomXOffset =
        (random.nextDouble() * screenWidth) - (screenWidth / 2);

    // 生成随机 y 偏移量（范围：-屏幕高度/2 到 屏幕高度/2）
    double randomYOffset =
        (random.nextDouble() * screenHeight) - (screenHeight / 2);

    // 计算最终随机坐标
    double randomX = centerX + randomXOffset;
    double randomY = centerY + randomYOffset;

    return Offset(randomX, randomY);
  }
}


lib/animation/animation_alpha_move_alpha.dart
import 'package:flutter/material.dart';

///动画  渐变-Y轴移动-渐变
class AnimationAlpahMoveAlpha extends StatelessWidget {
  late final Animation<double> alphaOut2;
  late final Animation<double> alphaIn2;
  late final Animation<Offset> moveUpAnim;
  final Offset startOffset, endOffset;
  //动画控制器
  late final Animation<double> controller;
  late final double offSetY;
  final Widget? childWidget;

  AnimationAlpahMoveAlpha(
      {super.key,
      required this.controller,
      required this.startOffset,
      required this.endOffset,
      this.offSetY = 50,
      required this.childWidget}) {
    //渐变出现0-1
    alphaIn2 = Tween<double>(
      begin: 0,
      end: 1.0,
    ).animate(
      CurvedAnimation(
        parent: controller,
        curve: const Interval(0.7, 0.74, curve: Curves.easeInOut),
      ),
    );

    /// 向上移动在指定的坐标endOffset位置向上Y轴移动offSetY距离
    moveUpAnim = Tween<Offset>(
      begin: endOffset,
      end: Offset(endOffset.dx, endOffset.dy - offSetY),
    ).animate(
      CurvedAnimation(
        parent: controller,
        curve: const Interval(0.74, 0.91, curve: Curves.easeInOut),
      ),
    );

    ///渐变消失 1-0
    alphaOut2 = Tween<double>(
      begin: 1,
      end: 0,
    ).animate(
      CurvedAnimation(
        parent: controller,
        curve: const Interval(0.91, 1, curve: Curves.easeInOut),
      ),
    );
  }

  ///动画展示的位置
  Widget _buildAnimation(BuildContext context, child) {
    return FadeTransition(
      opacity: alphaIn2,
      child: Transform.translate(
        offset: moveUpAnim.value,
        child: FadeTransition(
          opacity: alphaOut2,
          child: childWidget,
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: controller,
      builder: _buildAnimation,
    );
  }
}


lib/animation/animation_alpha_scale_move_alpha.dart
import 'package:flutter/material.dart';

///动画  渐变-缩放-移动-渐变
class AnimationAlphaScaleMoveAlpha extends StatelessWidget {
  //渐变展示——从0-1
  late final Animation<double> alphaIn;
  //渐变消失——1-0
  late final Animation<double> alphaOut;
  //缩放动画
  late final Animation<double> scaleAnim;
  //位移动画
  late final Animation<Offset> moveAnim;
  //起点坐标，终点坐标
  final Offset startOffset, endOffset;
  //动画控制器
  late final Animation<double> controller;
  //向上移动
  final double offSetY;
  //展示的widget
  final Widget? childWidget;
  AnimationAlphaScaleMoveAlpha(
      {super.key,
      required this.controller,
      required this.startOffset,
      required this.endOffset,
      required this.childWidget,
      this.offSetY = 50}) {
    //实现渐变展示
    alphaIn = Tween<double>(
      begin: 0,
      end: 1,
    ).animate(
      CurvedAnimation(
        parent: controller,
        //占总时间的
        curve: const Interval(0.0, 0.07, curve: Curves.easeInOut),
      ),
    );

    //缩放
    scaleAnim = Tween<double>(
      begin: 0,
      end: 1,
    ).animate(
      CurvedAnimation(
        parent: controller,
        //占总时间的
        curve: const Interval(0.0, 0.07, curve: Curves.easeInOut),
      ),
    );

    //停留在这里2300 0.69
    //移动
    moveAnim = Tween<Offset>(
      begin: startOffset,
      end: endOffset,
    ).animate(
      CurvedAnimation(
        parent: controller,
        curve: const Interval(0.5, 0.65, curve: Curves.easeInOut),
      ),
    );

    //渐变消失
    alphaOut = Tween<double>(
      begin: 1.0,
      end: 0.0,
    ).animate(
      CurvedAnimation(
        parent: controller,
        curve: const Interval(0.65, 0.7, curve: Curves.easeInOut),
      ),
    );
  }

  ///动画展示的位置
  Widget _buildAnimation(BuildContext context, child) {
    return FadeTransition(
      opacity: alphaIn,
      child: Transform.scale(
        scale: scaleAnim.value,
        child: Transform.translate(
          offset: moveAnim.value,
          child: Transform.scale(
            scale: alphaOut.value,
            child: childWidget,
          ),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: controller,
      builder: _buildAnimation,
    );
  }
}


lib/animation_demo.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter_animation_factory/tools/random_screen_offset.dart';

import 'audio_all_mic_gift.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    debugPaintBaselinesEnabled = true;
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const Demo(),
    );
  }
}

class Demo extends StatefulWidget {
  const Demo({super.key});

  @override
  State<StatefulWidget> createState() {
    return DemoState();
  }
}

class DemoState extends State<StatefulWidget> {
  List<Widget> widgetList = [];
  @override
  void initState() {
    super.initState();
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
  }

  _incrementCounter() async {
    // ignore: unused_local_variable
    // for (int i = 0; i < 100000; i++) {
    //   await Future.delayed(Duration(milliseconds: 150)); // 模拟异步操作
    //   setState(() {
    //     widgetList.add(AnimationFulWidget(
    //       endOffset: generateRandomScreenCoordinates(context),
    //       durationTime: 3000,
    //     ));
    //   });
    // }
    setState(() {
      widgetList.add(AudioAllMicGift(
        endOffset: RandomScreenOffset.generateRandomScreenCoordinates(context),
        durationTime: 3000,
        childWidget: const Icon(Icons.handshake),
        animationStateCallback: (status) {},
      ));
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("坐标动画"),
      ),
      body: Center(
        child: Stack(alignment: Alignment.center, children: widgetList),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
